core/src/main/scala/blobstore/Path.scala

```diff
@@ -15,67 +15,223 @@ Copyright 2018 LendUp Global, Inc.
-      path.key
-        .split(SEP)
-        .lastOption
-        .filterNot(_.isEmpty)
-        .map(s => if (path.isDir) s + SEP else s)
-        .getOrElse(path.root.split(SEP).lastOption.getOrElse("") + SEP)
-    }
+  /**
+    * Returns path with [[fileName]] field set to provided value.
+    * @param newFileName - value to be set as [[fileName]].
+    * @param reset - when true fields [[isDir]], [[size]] and [[lastModified]] in returned object would be set to None, otherwise kept the same.
+    * @return new Path with [[fileName]] field set to provided param.
+    */
+  def withFileName(newFileName: Option[String], reset: Boolean = true): Path =
+    if (newFileName == fileName) this
+    else
+      BasePath(
+        root,
+        pathFromRoot,
+        newFileName,
+        if (reset) None else size,
+        if (reset) None else isDir,
+        if (reset) None else lastModified
+      )
+
+  /**
+    * Returns path with [[size]] field set to provided value.
+    * @param newSize - value to be set as [[size]].
+    * @param reset - when true fields [[isDir]] and [[lastModified]] in returned object would be set to None, otherwise kept the same.
+    * @return new Path with [[size]] field set to provided param.
+    */
+  def withSize(newSize: Option[Long], reset: Boolean = true): Path =
+    if (newSize == size) this
+    else BasePath(root, pathFromRoot, fileName, newSize, if (reset) None else isDir, if (reset) None else lastModified)
+
+  /**
+    * Returns path with [[isDir]] field set to provided value.
+    * @param newIsDir - value to be set as [[isDir]].
+    * @param reset - when true fields [[size]] and [[lastModified]] in returned object would be set to None, otherwise kept the same.
+    * @return new Path with [[isDir]] field set to provided param.
+    */
+  def withIsDir(newIsDir: Option[Boolean], reset: Boolean = true): Path =
+    if (newIsDir == isDir) this
+    else BasePath(root, pathFromRoot, fileName, if (reset) None else size, newIsDir, if (reset) None else lastModified)
+
+  /**
+    * Returns path with [[lastModified]] field set to provided value.
+    * @param newLastModified - value to be set as [[lastModified]].
+    * @param reset - when true fields [[size]] and [[isDir]] in returned object would be set to None, otherwise kept the same.
+    * @return new Path with [[lastModified]] field set to provided param.
+    */
+  def withLastModified(newLastModified: Option[Instant], reset: Boolean = true): Path =
+    if (newLastModified == lastModified) this
+    else BasePath(root, pathFromRoot, fileName, if (reset) None else size, if (reset) None else isDir, newLastModified)
+
+  @SuppressWarnings(Array("scalafix:Disable.Any"))
+  override def equals(obj: Any): Boolean = obj match {
+    case p: Path =>
+      root == p.root &&
+        pathFromRoot == p.pathFromRoot &&
+        fileName == p.fileName &&
+        size == p.size &&
+        isDir == p.isDir &&
+        lastModified == p.lastModified
   }
```

Comment:
This is non-total, wonder why it's not captured by the compiler linters. Add a catch-all that returns false?